speechbrain/lobes/models/transformer/Conformer.py

```diff
@@ -99,20 +142,149 @@ def __init__(
+            applied_left_context = [
+                min(chunk_left_context, i * dct_config.chunk_size,)
+                for i in range(chunk_count)
+            ]
+
+            # build views of chunks with left context (but no 0-padding yet)
+            # the left context effectively becomes "left padding", we do not
+            # want to keep any convolution results centered on the left context
+            out = [
+                x[
+                    :,
+                    i * dct_config.chunk_size
+                    - applied_left_context[i] : (i + 1) * dct_config.chunk_size,
+                    ...,
+                ]
+                for i in range(chunk_count)
+            ]
+
+            out = [self.layer_norm(chk) for chk in out]
+            out = [chk.transpose(1, 2) for chk in out]
+            out = [self.bottleneck(chk) for chk in out]
+            out = [chk.transpose(1, 2) for chk in out]
+
+            # TODO: experiment around reflect padding, which is difficult
+            # because small chunks have too little time steps to reflect from
+            out = [
+                F.pad(
+                    out[i],
+                    (
+                        # channel dims, we do not to pad these
+                        0,
+                        0,
+                        # add missing left 0-padding if we lacked left context
+                        chunk_left_context - applied_left_context[i],
+                        # add missing right 0-padding as we disable default padding
+                        # also add missing frames of the rightmost chunk
+                        self.padding
+                        + (final_right_padding if i == len(out) - 1 else 0),
+                    ),
+                    value=0,
+                )
+                for i in range(len(out))
+            ]
+
+            # we pack together chunks in a single tensor so that we can feed it
+            # to the convolution directly. this is much more performant than
+            # doing the same with lists.
+
+            # -> [batch_size, num_chunks, chunk_size + lc + rpad, in_channels]
+            out = torch.stack(out, dim=1)
+
+            # -> [batch_size * num_chunks, chunk_size + lc + rpad, in_channels]
+            out = torch.flatten(out, end_dim=1)
+
+            # for the convolution:
+            # -> [batch_size * num_chunks, in_channels, chunk_size + lc + rpad]
+            out = out.transpose(1, 2)
+
+            # -> [batch_size * num_chunks, out_channels, chunk_size + rpad]
+            out = self._do_conv(out, inhibit_padding=True)
+
+            # -> [batch_size * num_chunks, chunk_size + rpad, out_channels]
+            out = out.transpose(1, 2)
+
+            out = self.after_conv(out)
+
+            # -> [batch_size, num_chunks, chunk_size, out_channels]
+            out = torch.unflatten(out, dim=0, sizes=(batch_size, -1))
+
```

Comment:
can't we do both in a single call?

Reply:
I'm not sure but IMO it is clearer this way (since we're "undoing" individual shape transforms) and should have no performance implication since this is only manipulating views and not doing concrete transforms in memory.